src/Spice86.Core/Emulator/InterruptHandlers/Dos/Ems/ExpandedMemoryManager.cs

```diff
@@ -0,0 +1,1236 @@
+    
+    public MemoryBlock MemoryBlock { get; }
+
+    public EmmMapping[] EmmSegmentMappings { get; } = new EmmMapping[64];
+
+    public EmmMapping[] EmmMappings { get; } = new EmmMapping[EmmHandle.EmmMaxPhysicalPages];
+    
+    public EmmHandle[] EmmHandles { get; } = new EmmHandle[EmmMaxHandles];
+    
+    public ushort MemorySizeInMb { get; }
+
+    public int TotalPages => MemoryBlock.Pages;
+
+    private readonly Ram _ram;
+
+    public ushort GetFreePages() => Math.Min((ushort) 0x7FFF, (ushort) (GetFreeMemoryTotal() / 4));
+
+    public ExpandedMemoryManager(Machine machine, ILoggerService loggerService) : base(machine, loggerService) {
+        _loggerService = loggerService.CreateLoggerConfiguration().MinimumLevel.Debug().CreateLogger().ForContext<ExpandedMemoryManager>();
+        MemorySizeInMb = Math.Max((ushort)8, (ushort) (_memory.Size / 1024 / 1024));
+        _ram = new Ram((uint) (MemorySizeInMb * 1024 * 1024));
+        var device = new CharacterDevice(DeviceAttributes.Ioctl, EmsIdentifier);
+        machine.Dos.AddDevice(device);
+        for (int i = 0; i < EmmHandles.Length; i++) {
+            EmmHandles[i] = new EmmHandle();
+        }
+
+        for (int i = 0; i < EmmMappings.Length; i++) {
+            EmmMappings[i] = new EmmMapping();
+        }
+
+        for (int i = 0; i < EmmSegmentMappings.Length; i++) {
+            EmmSegmentMappings[i] = new EmmMapping();
+        }
+
+        MemoryBlock = new MemoryBlock(MemorySizeInMb);
+
+        FillDispatchTable();
+        
+        _state.AH = EmmAllocateSystemHandle(8);
+    }
+    
+    /// <summary>
+    /// Allocates OS-dedicated handle (ems handle zero, 128kb) <br/>
+    /// This handle is never deallocated.
+    /// </summary>
+    /// <param name="logicalPage">The logical page number to allocate.</param>
+    /// <returns>Success or error code.</returns>
+    private byte EmmAllocateSystemHandle(ushort logicalPage) {
+        /* Check for enough free pages */
+        if ((GetFreeMemoryTotal() / 4) < logicalPage) {
+            if (_loggerService.IsEnabled(LogEventLevel.Warning)) {
+                _loggerService.Warning("EMS: Not enough free pages to allocate system handle. {FreePages}, {RequestedPages}", GetFreeMemoryTotal() / 4, logicalPage);
+            }
+            return EmmStatus.EmmOutOfLogicalPages;
+        }
+
+        /* Release memory if already allocated */
+        if (EmmHandles[EmmSystemHandle].LogicalPage != EmmNullHandle) {
+            ReleasePages(EmmHandles[EmmSystemHandle].MemHandle);
+        }
+        int mem = AllocatePages((ushort) (logicalPage * 4), false);
+        if (mem == 0) {
+            FailFastWithLogMessage("EMS: System handle memory allocation failure");
+        }
+        EmmHandles[EmmSystemHandle].LogicalPage = logicalPage;
+        EmmHandles[EmmSystemHandle].MemHandle = mem;
+        return EmmStatus.EmmNoError;
+    }
+
+    public uint Size => _ram.Size;
+    
+    public byte Read(uint address) {
```

Comment:
In Read() you want to figure out which physical page is being requested (0-4) and then figure out which logical page is mapped to that physical page at this moment. And then read from that logical page in _ram.

Or better yet, create a Ram for every new logical page and simply pass the read on to that. Then you don;t need the _ram at all and you can serve as many logical pages as you want.

Reply:
Thanks, I'm getting closer to this with my latest commit.